index.bs

```diff
@@ -1141,6 +1243,218 @@ steps:
+   failed:
+    1. If there were partial reads  and the number of modified bytes in |bytes|  is known,
+       set |result| to the number of modified bytes.
+    1. Else set |result| to 0.
+1. Let |arrayBuffer| be |buffer|'s [=underlying buffer=].
+1. [=write|Write=] |bytes| into |arrayBuffer|.
+1. Return |result|.
+
+</div>
+
+### The {{FileSystemSyncAccessHandle/write()}} method ### {#api-filesystemsyncaccesshandle-write}
+
+<div class="note domintro">
+  : |handle| . {{FileSystemSyncAccessHandle/write()|write}}(|buffer|, {{FilesystemReadWriteOptions}}: |options|)
+  :: Writes the content of |buffer| into the file associated with |handle| with |options|.{{FilesystemReadWriteOptions/at}} as the offset.
+</div>
+
+// TODO(fivedots): Is there a more formal way to describe the resutl of partial writes from the OS?
+
+// TODO(fivedots): Figure out mechanism to prevent concurrent IO operations e.g. doing write() while a truncate() is executing in parallel.
+<div algorithm>
+The <dfn method for=FileSystemSyncAccessHandle>write(|buffer|, {{FilesystemReadWriteOptions}}: |options|)</dfn> method, when invoked, must run
+these steps:
+
+1. If [=this=].[=[[state]]=] is "closed", throw a {{InvalidStateError}}.
+1. Let |writePosition| be |options|.{{FilesystemReadWriteOptions/at}}.
+1. Let |fileContents| be a copy of [=this=].[=[[file]]=]'s [=file entry/binary data=].
+1. Let |oldSize| be |fileContents|'s [=byte sequence/length=].
+1. Let |data| be [=get a copy of the buffer source|a copy of=] |buffer|.
+1. If |writePosition| is larger than |oldSize|,
+   append |writePosition| - |oldSize| `0x00` (NUL) bytes to the end of |fileContents|.
+
+   Note: Implementations are expected to behave as if the skipped over file contents
+   are indeed filled with NUL bytes. That doesn't mean these bytes have to actually be
+   written to disk and take up disk space. Instead most file systems support so called
+   sparse files, where these NUL bytes don't take up actual disk space.
+
+1. Let |head| be a [=byte sequence=] containing the first |writePosition| bytes of |fileContents|.
+1. Let |tail| be an empty [=byte sequence=].
+1. If |writePosition| + |data|.[=byte sequence/length=] is smaller than |oldSize|:
+   1. Let |tail| be a [=byte sequence=] containing the last
+      |oldSize| - (|writePosition| + |data|.[=byte sequence/length=]) bytes of |fileContents|.
+1. Let |newSize| be the [=byte sequence/length=] of the concatenation of |head|, |data| and |tail|.
+1. If |newSize| - |oldSize| exceeds the available [=storage quota=], throw a {{QuotaExceededError}}.
+1. Set [=this=].[=[[file]]=]'s [=file entry/binary data=] to the concatenation of |head|, |data| and |tail|.
+1. If the operations modifying the [=this=].[=[[file]]=]'s [=file entry/binary data=] in the previous steps
+   failed:
+    1. If there were partial modifications and the number of modified bytes is known,
+       return the number of modified bytes.
```

Comment:
**Important**

This is a bit vague since one reading is that the step above "modifies" the entire file. It would be good to make this more precise in a way that communicates exactly which number you're looking for. Probably it's how much of |data| was written?

Reply:
Done. Went for the number of written bytes from |data|.